File: apps/pwabuilder/Models/Android/AndroidPackageOptions.cs
namespace PWABuilder.Android.Models
{
    /// <summary>
    /// Options for creating an Android PWA package.
    /// </summary>
    public class AndroidPackageOptions
    {
        public string? AppVersion { get; set; }
        public int AppVersionCode { get; set; }
        public string? BackgroundColor { get; set; }
        public string? Display { get; set; }
        public bool EnableNotifications { get; set; }
        public bool? EnableSiteSettingsShortcut { get; set; }
        public string? FallbackType { get; set; }
        public AndroidFeatures? Features { get; set; }
        public string? Host { get; set; }
        public string? IconUrl { get; set; }
        public bool IncludeSourceCode { get; set; }
        public bool? IsChromeOSOnly { get; set; }
        public bool? IsMetaQuest { get; set; }
        public string? LauncherName { get; set; }
        public string? MaskableIconUrl { get; set; }
        public string? MonochromeIconUrl { get; set; }
        public string? Name { get; set; }
        public string? NavigationColor { get; set; }
        public string? NavigationColorDark { get; set; }
        public string? NavigationDividerColor { get; set; }
        public string? NavigationDividerColorDark { get; set; }
        public string? Orientation { get; set; }
        public string? PackageId { get; set; }
        public List<WebManifestShortcut>? Shortcuts { get; set; }
        public SigningOptions? Signing { get; set; }
        public string? SigningMode { get; set; } // "new", "none", "mine"
        public int SplashScreenFadeOutDuration { get; set; }
        public string? StartUrl { get; set; }
        public string? ThemeColor { get; set; }
        public string? ThemeColorDark { get; set; }
        public string? WebManifestUrl { get; set; }
        public string? PwaUrl { get; set; }
        public string? FullScopeUrl { get; set; }
        public int? MinSdkVersion { get; set; }

        public Validated Validate()
        {
            var errors = new List<string>();

            // Required fields (always mandatory)
            if (string.IsNullOrWhiteSpace(Name))
                errors.Add("Name is required.");
            if (string.IsNullOrWhiteSpace(PackageId))
                errors.Add("PackageId is required.");
            if (string.IsNullOrWhiteSpace(Host))
                errors.Add("Host is required.");
            if (string.IsNullOrWhiteSpace(PwaUrl))
                errors.Add("PwaUrl is required.");
            if (string.IsNullOrWhiteSpace(IconUrl))
                errors.Add("IconUrl is required.");
            if (string.IsNullOrWhiteSpace(LauncherName))
                errors.Add("LauncherName is required.");
            if (string.IsNullOrWhiteSpace(StartUrl))
                errors.Add("StartUrl is required.");
            if (string.IsNullOrWhiteSpace(BackgroundColor))
                errors.Add("BackgroundColor is required.");
            if (string.IsNullOrWhiteSpace(ThemeColor))
                errors.Add("ThemeColor is required.");
            if (string.IsNullOrWhiteSpace(Display))
                errors.Add("Display is required.");
            if (string.IsNullOrWhiteSpace(FallbackType))
                errors.Add("FallbackType is required.");
            if (string.IsNullOrWhiteSpace(WebManifestUrl))
                errors.Add("WebManifestUrl is required.");
            if (string.IsNullOrWhiteSpace(SigningMode))
                errors.Add("SigningMode is required.");

            if (IsMetaQuest == true && string.IsNullOrWhiteSpace(FullScopeUrl))
            {
                errors.Add("FullScopeUrl is required when IsMetaQuest is true.");
            }

            // Signing validation
            if (SigningMode != "none")
            {
                if (Signing == null)
                {
                    errors.Add("Signing is required when SigningMode is not 'none'.");
                }
                else
                {
                    if (SigningMode == "mine")
                    {
                        if (string.IsNullOrWhiteSpace(Signing.File))
                            errors.Add("Signing file is required in 'mine' mode.");
                        else if (!Signing.File.StartsWith("data:"))
                            errors.Add(
                                "Signing file must be base64 encoded (starting with 'data:')."
                            );
                    }

                    if (SigningMode == "new")
                    {
                        if (string.IsNullOrWhiteSpace(Signing.CountryCode))
                            errors.Add("CountryCode is required in 'new' mode.");
                        if (string.IsNullOrWhiteSpace(Signing.FullName))
                            errors.Add("FullName is required in 'new' mode.");
                        if (string.IsNullOrWhiteSpace(Signing.Organization))
                            errors.Add("Organization is required in 'new' mode.");
                        if (string.IsNullOrWhiteSpace(Signing.OrganizationalUnit))
                            errors.Add("OrganizationalUnit is required in 'new' mode.");
                    }

                    // Default passwords if missing
                    var defaultPass = Guid.NewGuid().ToString("N").Substring(0, 12);
                    Signing.KeyPassword ??= defaultPass;
                    Signing.StorePassword ??= defaultPass;

                    if (string.IsNullOrWhiteSpace(Signing.Alias))
                        errors.Add("Signing alias is required.");
                    if (string.IsNullOrWhiteSpace(Signing.KeyPassword))
                        errors.Add("Signing key password is required.");
                    if (string.IsNullOrWhiteSpace(Signing.StorePassword))
                        errors.Add("Signing store password is required.");
                }
            }

            if (errors.Any())
            {
                throw new ArgumentException(
                    "Invalid Android package options: " + string.Join(" | ", errors)
                );
            }

            // URI parsing
            var parsedHost = new Uri("https://" + Host!.Trim(), UriKind.Absolute);
            var parsedPwaUrl = new Uri(PwaUrl!.Trim(), UriKind.Absolute);
            var parsedIconUrl = new Uri(IconUrl!.Trim(), UriKind.Absolute);
            var parsedManifestUrl = new Uri(WebManifestUrl!.Trim(), UriKind.Absolute);
            Uri? parsedFullScope = !string.IsNullOrWhiteSpace(FullScopeUrl)
                ? new Uri(FullScopeUrl!.Trim(), UriKind.Absolute)
                : null;

            return new Validated(
                Name!.Trim(),
                PackageId!.Trim(),
                parsedHost,
                parsedPwaUrl,
                parsedIconUrl,
                LauncherName!,
                StartUrl!,
                BackgroundColor!,
                ThemeColor!,
                Display!,
                FallbackType!,
                SigningMode!,
                Signing!,
                parsedManifestUrl,
                parsedFullScope,
                AppVersion!,
                AppVersionCode,
                SplashScreenFadeOutDuration,
                IncludeSourceCode,
                MinSdkVersion ?? 19,
                IsChromeOSOnly ?? false,
                IsMetaQuest ?? false,
                NavigationColor,
                NavigationColorDark,
                NavigationDividerColor,
                NavigationDividerColorDark,
                Orientation,
                MaskableIconUrl,
                MonochromeIconUrl,
                EnableNotifications,
                EnableSiteSettingsShortcut ?? false,
                Features,
                Shortcuts
            );
        }
    }

    public class AndroidFeatures
    {
        public LocationDelegationConfig? LocationDelegation { get; set; }
        public PlayBillingConfig? PlayBilling { get; set; }
    }

    public class LocationDelegationConfig
    {
        public bool Enabled { get; set; }
    }

    public class PlayBillingConfig
    {
        public bool Enabled { get; set; }
    }

    public class WebManifestShortcut
    {
        public string? Name { get; set; }
        public string? Short_Name { get; set; }
        public string? Description { get; set; }
        public string? Url { get; set; }
        public List<ShortcutIcon>? Icons { get; set; }
    }

    public class ShortcutIcon
    {
        public string? Src { get; set; }
        public string? Sizes { get; set; }
    }

    public class SigningOptions
    {
        public string? File { get; set; }
        public string? Alias { get; set; }
        public string? FullName { get; set; }
        public string? Organization { get; set; }
        public string? OrganizationalUnit { get; set; }
        public string? CountryCode { get; set; }
        public string? KeyPassword { get; set; }
        public string? StorePassword { get; set; }
    }

    public record Validated(
        string Name,
        string PackageId,
        Uri Host,
        Uri PwaUrl,
        Uri IconUrl,
        string LauncherName,
        string StartUrl,
        string BackgroundColor,
        string ThemeColor,
        string Display,
        string FallbackType,
        string SigningMode,
        SigningOptions Signing,
        Uri ManifestUrl,
        Uri? FullScopeUrl,
        string AppVersion,
        int AppVersionCode,
        int SplashScreenFadeOutDuration,
        bool IncludeSourceCode,
        int MinSdkVersion,
        bool IsChromeOSOnly,
        bool IsMetaQuest,
        string? NavigationColor,
        string? NavigationColorDark,
        string? NavigationDividerColor,
        string? NavigationDividerColorDark,
        string? Orientation,
        string? MaskableIconUrl,
        string? MonochromeIconUrl,
        bool EnableNotifications,
        bool EnableSiteSettingsShortcut,
        AndroidFeatures? Features,
        List<WebManifestShortcut>? Shortcuts
    );
}

File: apps/pwabuilder-google-play/packaging/androidPackageOptions.ts
import { AppsFlyerConfig } from "@bubblewrap/core/dist/lib/features/AppsFlyerFeature.js";
import { FirstRunFlagConfig } from "@bubblewrap/core/dist/lib/features/FirstRunFlagFeature.js";
import { LocationDelegationConfig } from "@bubblewrap/core/dist/lib/features/LocationDelegationFeature.js";
import { PlayBillingConfig } from "@bubblewrap/core/dist/lib/features/PlayBillingFeature.js";
import { ShareTarget, WebManifestShortcutJson } from "@bubblewrap/core/dist/lib/types/WebManifest.js";
import { SigningOptions } from "./signingOptions.js";

/**
 * Options for generating the Android package.
 */
export type AndroidPackageOptions = {
    /**
     * Additional origins to be considered trusted by the app.
     */
    additionalTrustedOrigins?: string[],
    /**
     * The app version.
     */
    appVersion: string;
    /**
     * The app version code.
     */
    appVersionCode: number;
    /**
     * The app background color.
     */
    backgroundColor: string;
    /**
     * Display mode. 
     * standalone means the app displays like a normal app: with status bar and navbar visible. Recommended for most PWAs.
     * fullscreen means the app takes up every bit of available real-estate, no status bar or nav bar is visible. Recommended for games and immersive media experiences.
     * fullscreen-sticky is like fullscreen, except that when the user swipes from the device edge, the system bars appear semi-transparent, and the touch gesture is passed to your app so it can respond. Recommended for drawing apps and games that involve much swiping.
     */
    display: "standalone" | "fullscreen" | "fullscreen-sticky";
    /**
     * Whether to use Push Notification Delegation. If enabled, the TWA will be able to send push notifications without browser permission prompts.
     */
    enableNotifications: boolean;
    /**
     * Whether to show "site settings" / "app settings" shortcut for installed apps. https://twitter.com/daviddalbusco/status/1307033355020578818
     */
    enableSiteSettingsShortcut?: boolean;
    /**
     * Fallback behavior. "customtabs" = use Chrome's Custom Tabs feature as a fallback. "webview" = use a embedded web view as a fallback behavior.
     */
    fallbackType?: "customtabs" | "webview";
    /**
     * Configures AppsFlyer analytics.
     */
    features?: {
        appsFlyer?: AppsFlyerConfig;
        locationDelegation?: LocationDelegationConfig;
        playBilling?: PlayBillingConfig;
        firstRunFlag?: FirstRunFlagConfig;
    };
    /**
     * The URL host for the TWA, e.g. https://foo.com
     */
    host: string;
    /**
     * The URL to the icon to use for the app.
     */
    iconUrl: string;
    /**
     * Whether to include source code in the Android package.
     */
    includeSourceCode: boolean;
    /**
     * Setting to true will enable a feature that prevents non-ChromeOS devices.
     */
    isChromeOSOnly?: boolean;
    /**
     * Setting to true will make the app compatible with Meta Quest devices.
     */
    isMetaQuest?: boolean;
    /**
     * The name of the app used on the Android launch screen. This may be the same as name or may be a shortened version of that name to account for less available display space.
     */
    launcherName: string;
    /**
     * The URL to the maskable icon to use for the app.
     */
    maskableIconUrl?: string;
    /**
     * The URL to a monochrome icon to use for the app.
     */
    monochromeIconUrl?: string;
    /**
     * The Android app name.
     */
    name: string;
    /**
     * The color used for the navbar.
     */
    navigationColor: string;
    /**
     * The color used for the navbar in dark mode.
     */
    navigationColorDark?: string;
    /**
     * The color used for the navbar divider.
     */
    navigationDividerColor?: string;
    /**
     * The color used for the navbar divider in dark mode.
     */
    navigationDividerColorDark?: string;
    /**
     * 
     */
    orientation?: "default" | "any" | "natural" | "landscape" | "portrait" | "portrait-primary" | "portrait-secondary" | "landscape-primary" | "landscape-secondary";
    /**
     * The Android package ID to generate, e.g. com.mycompany.foo
     */
    packageId: string;
    /**
     * The URL to a service account JSON file used in communicating with Google Play APIs.
     */
    serviceAccountJsonFile?: string;
    /**
     * The share target from the web manifest.
     */
    shareTarget?: ShareTarget;
    /*
     * App shortcuts
     */
    shortcuts?: WebManifestShortcutJson[];
    /**
     * Details about the signing key. 
     * If .signingMode = "none", this will be ignored.
     * If .signingMode = "new", this must contain the signing details, but .signing.file will be ignored.
     * If .signingMode = "mine", this must contain the signing details and key file.
     */
    signing?: SigningOptions | null;
    /**
     * The signing operation to perform. "new" means create a new signing key. "none" means don't sign. "mine" means use the uploaded signing information.
     * If "new" in specified, then .signing must contain the details of the signing key.
     * If "mine" is specified, then .signing must contain the .signing.file and details of the signing key.
     */
    signingMode: "new" | "none" | "mine";
    /**
     * How long the splash screen should fade out in milliseconds
     */
    splashScreenFadeOutDuration: number;
    /**
     * The start url relative to the host.
     */
    startUrl: string;
    /**
     * The app theme color, e.g. #2f3d58. See https://developers.google.com/web/updates/2015/08/using-manifest-to-set-sitewide-theme-color
     */
    themeColor: string;
    /**
     * The app theme color dark.
     */
    themeColorDark?: string;
    /**
     * The URL to the web manifest.
     */
    webManifestUrl: string;
    /**
     * The navigation scope that the browser considers to be within the app. If the user navigates outside the scope, it reverts to a normal web page inside a browser tab or window. Must be a full URL. Required and used only by Meta Quest devices.
     */
    fullScopeUrl?: string;
    /**
     * The minimum [Android API Level](https://developer.android.com/guide/topics/manifest/uses-sdk-element#ApiLevels) required for the application to run. Defaults to `19`. Should be `23`, if `isMetaQuest` is `true`.
     */
    minSdkVersion?: number;
    /**
     * The URL of the PWA as input to pwabuilder.com
     */
    pwaUrl: string;
}